backend/bib_example_client.py

```diff
@@ -3,34 +3,66 @@
 
 import asyncio
 import grpc
+from datetime import datetime
 from example_bib_app.grpc import example_bib_app_pb2_grpc, example_bib_app_pb2
 
 
 async def main():
     async with grpc.aio.insecure_channel("localhost:50051") as channel:
-        stub = example_bib_app_pb2_grpc.AuthorControllerStub(channel)
-        
-        queue = asyncio.Queue()
-
-        async def generate_requests():
-            while True:
-                yield await queue.get()
-
-        await queue.put(input("List Authors ? (type enter to continue ....) \n"))
-
-        async for response in stub.List(generate_requests()):
-            print("Response received :", flush=True)
-            print(response.response, flush=True)
-            request = example_bib_app_pb2.AuthorListRequest()
-            await queue.put(request)
-
-        # async for response in stub.Stream(generate_requests()):
-        #     print("Response received :", flush=True)
-        #     print(response.response, flush=True)
-        #     request = example_bib_app_pb2.AuthorStreamRequest(
-        #         author_id=input("Give author id\n")
-        #     )
-        #     await queue.put(request)
+        # create Author gRPC client
+        author_client = example_bib_app_pb2_grpc.AuthorControllerStub(channel)
+        publisher_client = example_bib_app_pb2_grpc.PublisherControllerStub(channel)
+
+        book_client = example_bib_app_pb2_grpc.BookControllerStub(channel)
+
+        # Create
+        for i in range(10):
+            author_response = await author_client.Create(
+                example_bib_app_pb2.AuthorRequest(
+                    name_first="John",
+                    name_last=f"Doe{i}",
+                    birth_date=datetime.now().strftime("%Y-%m-%d"),
+                )
+            )
+            print("author create response:", author_response)  # flush=True
+
+        last_author_id = author_response.author_id
+
+        # Read
+        res = await author_client.List(example_bib_app_pb2.AuthorListRequest())
+        res.results
+        for author in res.results:
+            print(author.name_first, author.name_last)
+
+        # Update
+        author_response = await author_client.Update(
+            example_bib_app_pb2.AuthorRequest(
+                author_id=last_author_id,
+                name_first="Jane",
+                name_last="Doll",
+                birth_date=datetime.now().strftime("%Y-%m-%d"),
+            )
+        )
+        print("author update response:", author_response)  # flush=True
+
+        # Stream Read
+
+        async for book in  book_client.Stream(example_bib_app_pb2.BookStreamRequest()):
```

Comment:
Hi @AMontagu,
this line creates the error:
`  File "/lair/source/projects/science/django/django-socio-grpc-example/backend/bib_example_client.py", line 69, in <module>
    asyncio.run(main())
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/lair/source/projects/science/django/django-socio-grpc-example/backend/bib_example_client.py", line 50, in main
    async for book in  book_client.Stream(example_bib_app_pb2.BookStreamRequest()):
  File "/home/mark/py3venv/sociogrpc/lib/python3.10/site-packages/grpc/aio/_call.py", line 356, in _fetch_stream_responses
    await self._raise_for_status()
  File "/home/mark/py3venv/sociogrpc/lib/python3.10/site-packages/grpc/aio/_call.py", line 263, in _raise_for_status
    raise _create_rpc_error(
grpc.aio._call.AioRpcError: <AioRpcError of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "'InternalHttpRequest' object has no attribute 'write'"
        debug_error_string = "UNKNOWN:Error received from peer  {created_time:"2023-11-23T22:19:57.956831673+01:00", grpc_status:2, grpc_message:"\'InternalHttpRequest\' object has no attribute \'write\'"}"`

Reply:
do you know why ?

Reply:
I found the problem:
I accidentally unsed the sync server, rather than the async server:
 `./manage.py grpcrunaioserver --dev`